### PR TITLE
executor: reduce make slice for insert select statement

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: gengliqi <gengliqiii@gmail.com>

### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:
image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by you06 — DISMISSED** [review]

File: executor/insert_common.go [hunk 1]
```diff
@@ -778,18 +792,36 @@ func insertRowsFromSelectWorker(ctx context.Context, base insertCommon, batchSiz
...
+		rows = rows[:0]
+		rowsCacheIndex = 0
 		iter := chunk.NewIterator4Chunk(chk)
 		for innerChunkRow := iter.Begin(); innerChunkRow != iter.End(); innerChunkRow = iter.Next() {
-			innerRow := innerChunkRow.GetDatumRow(fields)
+			length := innerChunkRow.Len()
+			if len(datumRow) < length {
+				diff := length - len(datumRow)
+				for i := 0; i < diff; i++ {
+					datumRow = append(datumRow, types.Datum{}) //nolint:makezero
```
Comment (cfzjywxk) [inline executor/insert_common.go:803]: Is it necessary to set
```
for i := 0; i < len(datumRow); i++ {
    datumRow[i] = types.Datum{}
}
```
like the else branch?

File: executor/insert_common.go [hunk 2]
```diff
@@ -778,18 +792,36 @@ func insertRowsFromSelectWorker(ctx context.Context, base insertCommon, batchSiz
...
+			}
+			datumRow := innerChunkRow.GetDatumRowWithBuffer(fields, datumRow)
 			var row []types.Datum
-			row, err = e.getRow(ctx, innerRow)
+			if rowsCacheIndex >= len(rowsCache) {
+				rowsCache = append(rowsCache, make([]types.Datum, columnLen)) //nolint:makezero
+			}
+			row = rowsCache[rowsCacheIndex]
+			rowsCacheIndex++
+			row, err = e.getRowWithBuffer(ctx, datumRow, row, hasValue)
```
Comment (cfzjywxk) [inline executor/insert_common.go:818]: The generated row may use the byte reference in the `Datum` object which is easy to make mistakes, by now seems the cached `Datum` object is cleared each time so it's safe?

**Review by cfzjywxk — APPROVED** [review]
As it's quite urgent approve first.

File: executor/insert_common.go (hunk 1 above)
Comment (gengliqi) [reply executor/insert_common.go:803]: Addressed

File: executor/insert_common.go [hunk 3]
```diff
@@ -778,18 +792,36 @@ func insertRowsFromSelectWorker(ctx context.Context, base insertCommon, batchSiz
...
 			if err != nil {
 				logutil.Logger(ctx).Error("parallel insert get row error", zap.Error(err))
 				return err
 			}
 			if extraColsInSel != nil {
-				extraColsInSel = append(extraColsInSel, innerRow[e.rowLen:])
+				extraColsInSel = append(extraColsInSel, datumRow[e.rowLen:])
```
Comment (gengliqi) [inline executor/insert_common.go:824]: There is a bug that all `[]datum` in `extraColsInSel` are `datumRow`.

**Pull request merged**